Update README.md with more samples (#5)
akv-platform/setup-go: README.md

    uses: actions/cache@v3
    with:
      path: |
        ${{ env.cache }}
      key: setup-go-deps-${{ runner.os }}-go-${{ hashFiles('go.sum go.mod') }}

  - name:
    run: echo "$GOOS $GOARCH"> /tmp/env

  - name: Set up intermediate built files cache
    uses: actions/cache@v3
    with:
      path: |
        ${{ env.modcache }}
      key: setup-go-build-${{ env.GOOS }}-${{ env.GOARCH }}-${{ runner.os }}-go-${{ hashFiles('**/*.go /tmp/env') }}
      restore-keys: |
        setup-go-build-${{ env.GOOS }}-${{ env.GOARCH }}

```

### Restore-only caches
If there are several builds on the same repo it might make sense to create a cache in one build and use it in the
others. The action [actions/cache/restore](https://github.com/marketplace/actions/cache-restore)
should be used in this case.

### Generate different caches
This advanced use case assumes manual definition of cache key and requires the use of
[actions/cache](https://github.com/actions/cache/blob/main/examples.md#go---modules)

### Parallel builds
To avoid race conditions during the parallel builds they should either
[generate their own caches](#generate-different-caches), or create the cache

> for only one build and [restore](#restore-only-caches) that cache in the other builds.

Reviewer: ```suggestion
for only one build and [restore](#restore-only-caches) that cache is in other builds.
```

  - run: go run hello.go
  ```

**Caching in multirepos**
`cache-dependency-path` input assepts glob patterns and multi-line values:

```yaml
steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v4
    with:
      go-version: '1.17'
      cache-dependency-path: **/go.sum
  - run: go run hello.go
  ```

```yaml
steps:
  - uses: actions/checkout@v3
  - uses: actions/setup-go@v4
    with:
      go-version: '1.17'
      cache-dependency-path: |
        subdir1/go.sum
        subdir2/go.mod
  - run: go run hello.go

>   ```

Reviewer 2: I believe by multirepos you meant monorepos, such examples are already in the README as the PR https://github.com/actions/setup-go/pull/417 was recently merged. I think we can remove the highlighted lines.

Author: Actually, no. Monorepo stands for a project with single go.sum and multirepo stands for projects with few "roots".

I did rebase to actions/setup-go@main branch and combined the samples.

> ## Multi-target builds

Reviewer 2: ```suggestion
**Caching for multi-target builds**
```
Perhaps, there is no need to create a heading? Isn't bold text enough?

Author: I added 3rd level of heading in order to have link anchors for every sample

`cache-dependency-path` input used to generate unuque cache key and it is not limited to only
dependencies files. The common case is to add a file containing the extr info about the specific

> build, for example build target.

Reviewer 2: ```suggestion
The 'cache-dependency-path' input doesn't limit itself to dependency lock files only. It can also be used with additional files that contain details about the build settings. By using this method, caches for builds created for various operating systems, architectures, etc. can be separated.
```

```yaml
env:
  GOOS: ...
  GOARCH: ...

steps:

>   - run: echo "$GOOS $GOARCH"> /tmp/env

Reviewer 2: ```suggestion
  - run: echo "$GOOS $GOARCH" > /tmp/env
```
Not sure about the whitespace between the string and >. If It breaks the command, just ignore the comment.

  - uses: actions/setup-go@v4
    with:
      go-version: '1.17'

>       cache-dependency-path: go.sum /tmp/env

Reviewer 2: Is it tested to work with whitespace separated paths?

Author: it's confirmed glob the path does not work https://github.com/akv-demo/test-setup-go/actions/runs/6026013766/job/16348011349#step:4:74

Changes are made

  - run: go run hello.go    
```

> ## Invalidate cache when source code changes

Reviewer 2: ```suggestion
**Fine tuned caching the compiler's build outputs**
```

> Besides the dependencise the action caches the build outputs 
> ([GOCACHE](https://pkg.go.dev/cmd/go#hdr-Build_and_test_caching) directory) but by default this
> cache is not update in order to avoid unpredictable and frequent cache invaldation. Nevertheless
> including the source code files into `cache-dependency-path` input.

Reviewer 2: ```suggestion
Along with dependencies, the action also caches the build outputs from the compiler, specifically the [GOCACHE](https://pkg.go.dev/cmd/go#hdr-Build_and_test_caching) directory. By default, this part of the cache is only updated when the 'go.sum' file is modified. However, by entering source code file paths in the 'cache-dependency-path' input, you can override the default behavior and force the cache to be updated whenever a source file is changed.
```

```yaml
- uses: actions/setup-go@v4
  with:
    go-version: '1.17'
    cache-dependency-path: go.sum **/*.go
- run: go run hello.go
```

But more practically to manage the cache with the text file manually updated according to the amount
of changes made in the repo.

```yaml
- uses: actions/setup-go@v4
  with:
    go-version: '1.17'
    cache-dependency-path: go.sum cache-version.txt
- run: go run hello.go

> ```

Reviewer 2: Frankly, I don't see this option as more practical than the option above, I'd remove it.

Author: I removed it, but still suspect it makes sense to show a way to reduce the frequency of the cache updates

If the nvaldation happens too often it make to manage the cache with the text file
manually updated according to the amount of changes made in the repo.

```yaml
- uses: actions/setup-go@v4
  with:
    go-version: '1.17'
    cache-dependency-path: go.sum cache-version.txt
- run: go run hello.go
```

> ## Caching with actions/cache

Reviewer 2: ```suggestion
**Using actions/cache action**
```

> The caching capabilities of the action are limited for the simplest builds and can be ineffective in the real world
> use cases. If the build requires fine-grained turning the built-in caching should be disabled and
> [actions/cache](https://github.com/actions/cache) should be used.

Reviewer 2: ```suggestion
The caching capabilities of setup-go are limited to the simplest and most popular use cases. If fine-grained tuning of caching is required, it's recommended to disable the built-in caching and use [actions/cache](https://github.com/actions/cache).
```

> Here the sample `actions/cache` configuration with the extending options allowing 
>  - configuring paths to cache
>  - have different caches for rare changes dependencies and more often changed intermediate build files
>  - use `restore-key` input to restore the previous cache even if the current key cache has changed
>  - have different caches intermediate build files of different architectures
>  - have custom cache key for parallel builds

Reviewer 2: ```suggestion
The example workflow below utilizes the `actions/cache` action and adds flexibility, for instance, it:
 - allows to configure cache path
 - allows to have different caches for rare changed dependencies and more often changed intermediate build files
 - uses the `restore-key` input to restore the previous cache even if the current key cache has changed
 - has different caches for the compiler's build outputs for different architectures
 - has custom cache keys for parallel builds
```

```yaml
build:
  env:
   GOOS: ...
   GOARCH: ...

  steps:
  - uses: actions/setup-go@v4
    with:
      go-version: "1.17"
      cache: false

  - name: Get Go cached paths
    run: |
      echo "cache=$(go env GOCACHE)" >> $GITHUB_ENV
      echo "modcache=$(go env GOMODCACHE)" >> $GITHUB_ENV

  - name: Set up dependencies cache
    uses: actions/cache@v3
    with:
      path: |
        ${{ env.cache }}
      key: setup-go-deps-${{ runner.os }}-go-${{ hashFiles('go.sum go.mod') }}
      restore-keys: |

>         setup-go-deps-${{ runner.os }}-go-

Reviewer 2: I'm not sure about advising customers to use restore keys for GO dependencies, because of the `dependencies snowball` effect.

  - name:
    run: echo "$GOOS $GOARCH"> /tmp/env

  - name: Set up intermediate built files cache
    uses: actions/cache@v3
    with:
      path: |
        ${{ env.modcache }}
      key: setup-go-build-${{ env.GOOS }}-${{ env.GOARCH }}-${{ runner.os }}-go-${{ hashFiles('**/*.go /tmp/env') }}
      restore-keys: |
        setup-go-build-${{ env.GOOS }}-${{ env.GOARCH }}

```

## Restore-only caches
If there are several builds on the same repo it might make sense to create a cache in one build and use it in the
others. The action [actions/cache/restore](https://github.com/marketplace/actions/cache-restore)
should be used in this case.

## Generate different caches
This advanced use case assumes manual definition of cache key and requires the use of
[actions/cache](https://github.com/actions/cache/blob/main/examples.md#go---modules)

## Parallel builds
To avoid race conditions during the parallel builds they should either
[generate their own caches](#generate-different-caches), or create the cache
for only one build and [restore](#restore-only-caches) that cache in the other builds.

> 

Reviewer 2: I tried to investigate https://github.com/marketplace/actions/cache-restore action and it seems it's not that popular, why do you decide to suggest it to the customers? It's not even from the actions org.

Author: It was requested here https://github.com/actions/setup-go/issues/316

Use case is workflow that only consumes the results of another build: i.e. tests build. They must not store their intermediate files.
